Update matlab bindings adding inverse-kinematics

lrapetti opened the pull request:
With this PR bindings for `inverse-kinematics` are generated.

The commit has been done accordingly to the [guidelines](https://github.com/robotology/idyntree/blob/master/doc/generating-idyntree-matlab-bindings.md).

@traversaro is it ok to target `master` with the updated bindings or should `devel` be targeted instead?

traversaro commented:
> @traversaro is it ok to target `master` with the updated bindings or should `devel` be targeted instead?

Let's target `devel` as this is a new feature, thanks. I cleaned up the release process for 1.0, so I think a 1.1 release should be faster.

traversaro commented:
Can you also update the Changelog at https://github.com/robotology/idyntree/blob/devel/CHANGELOG.md ? Thanks!

lrapetti commented:
Sorry, by mistake I committed also the change in https://github.com/robotology/idyntree/pull/631, I'll clean it up

lrapetti commented:
> Sorry, by mistake I committed also the change in #631, I'll clean it up

Fixed

> Let's target `devel` as this is a new feature, thanks. I cleaned up the release process for 1.0, so I think a 1.1 release should be faster.

Target changed to `devel`

> Can you also update the Changelog at https://github.com/robotology/idyntree/blob/devel/CHANGELOG.md ? Thanks!

Done https://github.com/robotology/idyntree/pull/633/commits/d7e2035dce1422c72005277f0d9df9f8c21dd16a

traversaro commented on CHANGELOG.md:
```diff
@@ -9,6 +9,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+### Added
+- Added `bindings` for `InverseKinematics` (https://github.com/robotology/idyntree/pull/633)
+
```
Sorry, probably editing the ChangeLog is still confusing. The release 1.0.1 has been already done, so this change will go in a currently `unreleased` version, so it should be added in the `## [Unreleased]` section, see https://keepachangelog.com/en/1.0.0/ for more details.

lrapetti replied:
ah ok! Yes, actually in fact I was a bit confused and I was not sure about the position

lrapetti replied:
It is now moved under `Unreleased`

traversaro commented:
I just realized that unfortunately we can't merge the PR in its current form, sorry.

The main problem is that as it is now, we are breaking the following configuration (that is used, especially in Windows, in some repos, even if it is not tested by iDynTree's repo CI): 
* `IDYNTREE_USES_IPOPT = OFF`
* `IDYNTREE_USES_MATLAB = ON` or `IDYNTREE_USES_OCTAVE = ON`

This is due to the fact that we commit the generated source code with the modified SWIG that supports matlab.
We faced this problem in the past also with other dependencies, in particular Irrlicht for the visualization library and Assimp for the solidshapes helpers library. In both cases, the solution was to compile a dummy version of the related libraries when the specified dependency is missing, that if used created an error at runtime, see: 
* https://github.com/robotology/idyntree/blob/v1.0.1/src/visualization/src/Visualizer.cpp#L88
* https://github.com/robotology/idyntree/blob/v1.0.1/src/visualization/CMakeLists.txt#L5
* https://github.com/robotology/idyntree/blob/v1.0.1/src/solid-shapes/src/InertialParametersSolidShapesHelpers.cpp#L400

In the  case of InverseKinematics, I think the easiest option is to prepare a dummy  InverseKinematics.cpp to compile when Ipopt is not available https://github.com/robotology/idyntree/blob/v1.0.1/src/inverse-kinematics/src/InverseKinematics.cpp , that in place of all the methods just prints an error that explains that `IDYNTREE_USES_IPOPT` should be enabled to use that class. 
Sorry for the boring work, but without this or something equivalent it would be complicate to me to merge such a change.

lrapetti commented:
> In the case of InverseKinematics, I think the easiest option is to prepare a dummy InverseKinematics.cpp to compile when Ipopt is not available https://github.com/robotology/idyntree/blob/v1.0.1/src/inverse-kinematics/src/InverseKinematics.cpp , that in place of all the methods just prints an error that explains that `IDYNTREE_USES_IPOPT` should be enabled to use that class.
> Sorry for the boring work, but without this or something equivalent it would be complicate to me to merge such a change.

Ok, I can work this out. Does the "dummy class" need to implement all the actual methods of the class?

traversaro commented:
>  Ok no worry, I can work this out. Does the "dummy class" need to implement all the actual methods of the class?

Yes, otherwise you would get a linking error.

lrapetti commented:
With https://github.com/robotology/idyntree/pull/633/commits/0657826807f4d2840c984fbc859326d33648e2ed I have implemented the dummy methods. However I was thinking that a possible cleaner solutions could be:
-  to define directly a dummy class [`InverseKinematicsDataDummy`](https://github.com/robotology/idyntree/blob/master/src/inverse-kinematics/src/InverseKinematics.cpp#L26) so we would move almost all the conditions in that class. So eventually in the future we may have a different solver working for windows and we can implemente the class for the other solver.
- to use just a single `if` redefining all the methods. But I went for using multiple ifs because in this case if a new method is add this avoid forgetting adding it also in the NOT_USE_ASSIMP case.

ahoarau commented on src/CMakeLists.txt:
```diff
@@ -20,10 +20,10 @@ add_subdirectory(model_io)
-  add_subdirectory(inverse-kinematics)
 endif()
 
 if (IDYNTREE_COMPILES_OPTIMALCONTROL)
```
IDYNTREE_USES_OPTIMALCONTROL instead ?

lrapetti replied:
I think currently the way the option are used is: `IDYNTREE_USES_{dependency_name}` (e.g. for `IDYNTREE_USES_YARP` or `IDYNTREE_USES_IPOPT`), and `IDYNTREE_COMPILES_{IDynTree component}` (e.g. [`IDYNTREE_COMPILE_TESTS`](https://github.com/robotology/idyntree/blob/master/src/CMakeLists.txt#L12)). So in this case I would say `IDYNTREE_COMPILES_OPTIMALCONTROL` is coherent with the rest of the code, even we might want to change and use always `USE`.

Anyway, this part did not change within this PR, so eventually the discussion might be moved elsewhere.

traversaro commented:
> But I went for using multiple ifs because in this case if a new method is add this avoid forgetting adding it also in the NOT_USE_ASSIMP case.

Yes, this make sense, thanks!

traversaro commented on src/inverse-kinematics/CMakeLists.txt:
```diff
@@ -37,6 +37,7 @@ target_include_directories(${libraryname} SYSTEM PRIVATE ${EIGEN3_INCLUDE_DIR})
 target_link_libraries(${libraryname} idyntree-core idyntree-high-level ${IPOPT_LIBRARIES})
 
 target_compile_definitions(${libraryname} PRIVATE ${IPOPT_DEFINITIONS})
+target_compile_definitions(${libraryname} PRIVATE IDYNTREE_USES_IPOPT)
```
As you can see from the GitHub Actions failures, you need to also modify the CMakeLists.txt of this library to link IPOPT, pass the `IDYNTREE_USES_IPOPT` definition, add the private sources that depend on IPOPT only if `IDYNTREE_USES_IPOPT ` is set to FALSE, similarly to what is done for assimp in the solid-shapes lib, see https://github.com/robotology/idyntree/blob/v1.0.1/src/solid-shapes/CMakeLists.txt . Note that as we require CMake 3.5 now, if necessary you can also use `target_sources` command, see https://github.com/robotology/idyntree/blob/v1.0.1/src/solid-shapes/CMakeLists.txt .

traversaro commented:
See GitHub Actions failures and https://github.com/robotology/idyntree/pull/633#discussion_r368252019 .

lrapetti commented:
> See GitHub Actions failures and [#633 (comment)](https://github.com/robotology/idyntree/pull/633#discussion_r368252019) .

it should be fixed in https://github.com/robotology/idyntree/pull/633/commits/ee3d335ac21c39d724c3846ee4fdefe641c89ea6.

lrapetti commented on src/inverse-kinematics/CMakeLists.txt:
```diff
@@ -11,18 +11,22 @@ set(IDYN_TREE_IK_HEADERS include/iDynTree/ConvexHullHelpers.h
-
-add_library(${libraryname} ${IDYN_TREE_IK_HEADERS} ${IDYN_TREE_IK_SOURCES}
-                           ${PRIVATE_IDYN_TREE_IK_SOURCES} ${PRIVATE_IDYN_TREE_IK_HEADERS})
+if(IDYNTREE_USES_IPOPT)
```
I had this `if` since those classes are not compiled in the dummy class.
@traversaro please check if it looks sound for you or if you think there is a better way to handle it.

traversaro replied:
Using `target_sources` you can avoid to duplicate the `add_library` call, in any case what you did is ok, thanks!

lrapetti commented:
Some tests, apparently not related to inverse-kinematics, are failing

traversaro commented:
> 
> 
> Some tests, apparently not related to inverse-kinematics, are failing

Yes, those are probably not related to the PR and caused by https://github.com/robotology/yarp/issues/2189 .

traversaro commented:
> Some tests, apparently not related to inverse-kinematics, are failing

Failures are not related to PR, but it seems to be related to  https://github.com/robotology/yarp/issues/2189 .

traversaro commented:
Can you rebase all the commits except https://github.com/robotology/idyntree/pull/633/commits/ea370f344e27fcf6c7ca1218ed4c7459e7004b13 or in genear clean the history? After that, the change can be merged.

lrapetti commented:
> Can you rebase all the commits except [ea370f3](https://github.com/robotology/idyntree/commit/ea370f344e27fcf6c7ca1218ed4c7459e7004b13) or in genear clean the history? After that, the change can be merged.

Done

traversaro commented:
> > Can you rebase all the commits except [ea370f3](https://github.com/robotology/idyntree/commit/ea370f344e27fcf6c7ca1218ed4c7459e7004b13) or in genear clean the history? After that, the change can be merged.
> 
> Done

Thanks!
